### ISSUE TITLE
The behavior of `torch.einsum` significantly differs between TPU and other devices.

### ISSUE THREAD

**Issue by jqhoogland** [pytorch/xla#7050]
## 🐛 Bug

Learning dynamics systematically differ between TPU and CPU/MPS/CUDA if the computation graph includes `torch.einsum`. See the loss traces pictured below ([the increase in loss is because I'm adding a bit of noise to the parameters between updates](https://en.wikipedia.org/wiki/Stochastic_gradient_Langevin_dynamics)): 

![image](https://github.com/pytorch/xla/assets/28595964/354094e9-ac76-4780-8ea8-4e5bed403540)

Figure: Loss traces obtained by running stochastic gradient Langevin dynamics (SGLD = SGD + Gaussian noise) on a transformer trained on a natural language task. Depicted are two implementations of the transformer (one with einsums & one without) and two different devices. On CPU, the two implementations lead to exactly the same behavior. On TPUs, the two implementations lead to significantly different behavior — in particular, the combination of TPUs & einsum implementation diverges from the other traces. This is consistent across seeds, datasets, and at least two different architectures. 

## To Reproduce

See [this notebook](https://colab.research.google.com/drive/1ozujifl3CgTvcTnwRLDfziYulgsJG1En?usp=sharing) used to generate the above figure. 

Unfortunately, the notebook isn't quite as minimal as I would like: 
- The discrepancy was harder to show clearly on a synthetic dataset (so I'm using a natural language dataset)
- I'm using a transformer implementation provided by [transformer_lens](https://github.com/neelnanda-io/TransformerLens). The notebook includes a fork of this implementation that avoids the use of einsum. 
- I'm using a custom optimizer to implement [SGLD](https://en.wikipedia.org/wiki/Stochastic_gradient_Langevin_dynamics). I think this is a necessary ingredient. 
- It takes a few minutes to run through (the CPU runs are especially slow).
- This notebook doesn't work on Colab (or Kaggle) but gets stuck hanging. I think this is a problem for another issue. 

## Expected behavior

I expect the loss traces to resemble what I get on MPS, CPU, & CUDA. See figures above. 

## Environment

 - Python 3.8
 - Reproducible on XLA backend [CPU/TPU/CUDA]: TPU
 - torch_xla version: 2.2 & 2.3

## Additional context

I was working on implementing [SGLD](https://en.wikipedia.org/wiki/Stochastic_gradient_Langevin_dynamics)-based posterior sampling for [local learning coefficient estimation](https://arxiv.org/abs/2308.12108) on TPUs, building on [a library I maintain](https://github.com/timaeus-research/devinterp). 

In less jargon-y terms, the aim is "to probe the local geometry of the loss landscape". In order to do that, I sample a bunch of points in the neighborhood of some starting choice of weights `w_star`. Sampling these points amounts to running SGD-based training starting at `w_star` with two modifications: (1) add a bit of Gaussian noise between optimization steps, and (2) add a "localizing term" to keep the samples close to `w_star` (=L2 regularization, but centered at `w_star` rather than at 0). The details of this aren't relevant (you can see the implementation [here](https://github.com/timaeus-research/devinterp/blob/202da47182bc38c8b42c05ced17741194e2ac751/src/devinterp/optim/sgld.py) if you want). 

The important thing is that the thing I want to measure is the average loss over all of these local samples. What I noticed was that this average was consistent across MPS, CPU, and CUDA, but systematically differs between those devices and TPUs. 

As I investigated, I realized the problem had to do with my use of a transformer implementation provided by [transformer_lens](https://github.com/neelnanda-io/TransformerLens). When I used implementations provided by HF, I noticed no difference across devices. When I forked the TransformerLens implementation and replaced every instance of torch.einsum and fancy_einsum with native torch functionality, I eliminated the difference between these devices. 

From this, I produced the minimal notebook above. It's still not quite as minimal as I would like, but it should be enough to help get started.

cc @alanwaketan

**Comment by JackCaoG** [comment]
Hey @jqhoogland thanks for the detail analysis. It seems like issue is with how we lower `torch.einsum` vs normla matmul. Can you provide a repo similar to 
```
import torch
import torch_xla
import torch_xla.core.xla_model as xm


device = torch_xla.device()

x = torch.randn(5).to(device)
y = torch.randn(4).to(device)

# force a execution to make usre all other tensors on XLA device are datas instead of pending IRs
xm.mark_step()
res = torch.einsum('i,j->ij', x, y)
print(torch_xla._XLAC._get_xla_tensors_hlo([res]))
```
which prints something like
```
HloModule IrToHlo.6, entry_computation_layout={(f32[4]{0}, f32[5]{0})->(f32[5,4]{1,0})}

ENTRY %IrToHlo.6 (p0.1: f32[4], p1.2: f32[5]) -> (f32[5,4]) {
  %p1.2 = f32[5]{0} parameter(1)
  %p0.1 = f32[4]{0} parameter(0)
  %dot.3 = f32[5,4]{1,0} dot(f32[5]{0} %p1.2, f32[4]{0} %p0.1), lhs_contracting_dims={}, rhs_contracting_dims={}, frontend_attributes={grad_x="false",grad_y="false"}
  %transpose.4 = f32[5,4]{1,0} transpose(f32[5,4]{1,0} %dot.3), dimensions={0,1}
  ROOT %tuple.5 = (f32[5,4]{1,0}) tuple(f32[5,4]{1,0} %transpose.4)
}
```

You can do this with non-einsum so we can compare the HLO. If you also want to compare the backward you will can do something similar to 
```
print(torch_xla._XLAC._get_xla_tensors_hlo([res, x.grad, y.grad]))
```
which will print the graph to compute the result and gradient of your choice.

**Comment by ysiraichi** [comment]
Closing this due to inactivity.
